bottlerocket-os/bottlerocket
# Integrate `systemd-resolved` when using `systemd-networkd`

**Issue number:**
Related to #2449 

**Description of changes:**
This PR is one of, if not the final pieces of the puzzle to integrate `systemd-networkd` into Bottlerocket!

The set of commits here represent a slight change in direction for DNS resolution when using `systemd-networkd`.  Rather than use a custom DBUS listener ([whippet](#3270)) to react to interface changes and write a `resolv.conf`, we can rely on `systemd-resolved` to handle that.  These commits represent the steps needed to integrate `systemd-resolved` into Bottlerocket.

The first few commits bring in the `systemd-resolved` package and integrate it into the OS when using `systemd-resolved`.  There are a few additional mount units added, for directories needed early in boot for configuration files/drop-ins.

The next notable change (which can be updated when `wicked` is removed), is kubelet's configured `resolveConf` path.  Since we can't conditionalize this path based on the network backend, it now points at a symlink that `netdog` will manage and write early in boot.  For reasons described in [this k8s documentation](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/#known-issues) we are unable to continue to point kubelet at `/etc/resolv.conf` when using **`systemd-resolved`.**
```
bash-5.1# ls -lah /run/netdog/                                                                                                                      
...                                                                                  
lrwxrwxrwx.  1 root root  32 Aug 25 19:40 resolv.conf -> /run/systemd/resolve/resolv.conf 
```

Another notable change is the addition of a oneshot systemd unit that runs `netdog write-network-status` after the network is up, but before other services.  This unit functions similarly as the existing `netdog install` hook that wicked calls, writing the current IP to file as well as handling the aforementioned `resolv.conf` symlink.  Since `systemd-networkd` does not provide hooks into the interface bringup process, we use this oneshot unit instead.

Finally, the last significant change (which probably contains the most code changes) is the handling of custom API settings for DNS name servers and search domains.  Bottlerocket calls a "restart command" `netdog write-resolv-conf`  every single boot when generating DNS configuration files, and calls the same command each time those settings change.  In this PR, that code has been updated when using `systemd-networkd`.  If there aren't any custom DNS settings, `netdog` does nothing and leaves all DNS handling to `systemd-resolved`, who gets DNS information from the DHCP lease.

If there are custom DNS settings, `netdog` will write drop-ins for each configured interface. 
 The drop-in will turn off the `UseDNS` or `UseDomains` setting for the interface, depending on the custom DNS settings.  If, for example, there are custom search domain settings, the drop-in will contain `UseDomains=false` for the interface, meaning it won't use domains acquired via DHCP lease.  Finally, `netdog` will drop a drop-in for `systemd-resolved`, using the custom DNS setting.
 
 An example of the interface drop-in when setting custom name servers and the corresponding `resolvectl` output showing overrides:
 ```
bash-5.1# cat /etc/systemd/network/10-eth0.network
[Match]
Name=eth0
[Link]
RequiredForOnline=true
RequiredFamilyForOnline=ipv4
[Network]
DHCP=yes
IPv6AcceptRA=true
IPv6DuplicateAddressDetection=0
[DHCPv4]
UseMTU=true
[DHCPv6]
WithoutRA=solicit
[IPv6AcceptRA]
UseMTU=true

bash-5.1# cat /etc/systemd/network/10-eth0.network.d/10-dns.conf 
[Network]
DNSDefaultRoute=false
[DHCPv4]
UseDNS=false
UseDomains=true
[DHCPv6]
UseDNS=false
UseDomains=true
[IPv6AcceptRA]
UseDNS=false
UseDomains=true

bash-5.1# resolvectl                                                             
Global
         Protocols: -LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
  resolv.conf mode: stub
Current DNS Server: 1.2.3.4
        DNS Servers 1.2.3.4 192.168.0.2

Link 2 (eth0)
Current Scopes: none
     Protocols: -DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
    DNS Domain: us-west-2.compute.internal
...

bash-5.1# cat /etc/systemd/resolved.conf.d/10-resolv.conf 
[Resolve]
DNS=1.2.3.4 192.168.0.2
```

**Testing done:**
* Boot an aws-k8s-1.24 instance *using systemd-networkd* and validate the following:
  * The instance comes up and joins the cluster properly and can run pods
  * The resolveConf symlink is correct `resolv.conf -> /run/systemd/resolve/resolv.conf`
  * Set custom DNS settings for _just_ domains and servers, and also both, and observe the proper drop-ins being written for each interface as well as `systemd-resolved`
  * Ensure pods are able to be run before and after setting DNS settings via API (set to the exact same values) in order to test that the global resolved DNS information is used
* Boot an aws-k8s-1.24 instance *using wicked* and validate the same as above, except making sure that custom DNS settings end up in `/etc/resolv.conf`
  * Ensure `libnss_resolve.so` is not present on an instance using wicked and host resolution works ok with the new `nsswitch.conf`
 
* Following the ["Debugging DNS Resolution"](https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/#known-issues) instructions in the k8s docs, ensure DNS resolution is happy:
```
$ kubectl exec -i -t dnsutils -- nslookup kubernetes.default
Server:         10.100.0.10
Address:        10.100.0.10#53

Name:   kubernetes.default.svc.cluster.local
Address: 10.100.0.1
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
